src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs

```diff
@@ -469,5 +469,97 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
 
             return false;
         }
+
+        public static bool IsInInactiveRegion(
+            this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(syntaxTree);
+
+            // cases:
+            // $ is EOF
+
+            // #if false
+            //    |
+
+            // #if false
+            //    |$
+
+            // #if false
+            // |
+
+            // #if false
+            // |$
+
+            if (syntaxTree.IsPreProcessorKeywordContext(position, cancellationToken))
+            {
+                return false;
+            }
+
+            // The latter two are the hard cases we don't actually have an 
+            // DisabledTextTrivia yet. 
+            var trivia = syntaxTree.GetRoot(cancellationToken).FindTrivia(position, findInsideTrivia: false);
+            if (trivia.Kind() == SyntaxKind.DisabledTextTrivia)
+            {
+                return true;
+            }
+
+            var token = syntaxTree.FindTokenOrEndToken(position, cancellationToken);
+            if (token.Kind() == SyntaxKind.EndOfFileToken)
+            {
+                var triviaList = token.LeadingTrivia;
+                foreach (var triviaTok in triviaList.Reverse())
+                {
+                    if (triviaTok.Span.Contains(position))
+                    {
+                        return false;
+                    }
+
+                    if (triviaTok.Span.End < position)
+                    {
+                        if (!triviaTok.HasStructure)
+                        {
+                            return false;
+                        }
+
+                        var structure = triviaTok.GetStructure();
+                        if (structure is BranchingDirectiveTriviaSyntax branch)
+                        {
+                            return !branch.IsActive || !branch.BranchTaken;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public static bool IsPreProcessorKeywordContext(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            return IsPreProcessorKeywordContext(
+                syntaxTree, position,
+                syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken, includeDirectives: true));
+        }
+
+#pragma warning disable IDE0060 // Remove unused parameter
+        public static bool IsPreProcessorKeywordContext(this SyntaxTree syntaxTree, int position, SyntaxToken preProcessorTokenOnLeftOfPosition)
```

Comment:
I understand this a move, but seems this can be an extension method on the token instead of the unused `syntaxTree` parameter.

Reply:
yes.  but that would have a downstream impact on all the existing callers.  i didn't want to do that.  :)

Reply:
Sounds reasonable.